### PR TITLE
[WIP] Devirtualizing vtable method calls

### PR DESCRIPTION
**Pull request by HoundThe**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md#code-style)
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [radare2 book](https://github.com/radareorg/radare2book) with the relevant information (if needed)

**Detailed description**

First part of #12603 
Implemented command that takes vtable of a class, and finds actual destination of a method on certain offset in the vtable. I used code already made for vtables in vtables.c to handle reading and endiannes, not sure if this is the correct way and if I didn't overlook something.

this is the example input and output. (I've made a simple c++ binary with class B and C, each of them have their own vtable with function "bar" and "qux", confirmed the addresses are correct from the disassembly)
```
[0x000010e0]> av

Vtable Found at 0x00003d30
0x00003d30 : method.C.bar
0x00003d38 : method.B.qux


Vtable Found at 0x00003d50
0x00003d50 : method.B.bar
0x00003d58 : method.B.qux

[0x000010e0]> ac B
[0x000010e0]> acv B 0x3d50 0x0 0x10
[0x000010e0]> acv B
B:
     0 vtable 0x3d50 @ +0x0 size:+0x10
[0x000010e0]> acvf 0x0 B
Function address: 0x000011ca
[0x000010e0]> acvf 0x8 B
Function address: 0x00001206
[0x000010e0]> 
```

### CODE REVIEW

**Comment by HoundThe** [comment]
Changed the naming and added ability to list all possible options in all vtables 
example input/output rn

```
av
ac B
ac C
acv B 0x3d50 0x0 0x10
acv B 0x3d50 0x0 0x8
acv C 0x00003d30 0x0 0x10
acv B
acll
!echo "Printing functions on offset 0x0 and 0x8 in B vtable"
acvf 0x0 B
acvf 0x08 B
!echo "Printing functions on offset 0x0 and 0x8 in C vtable"
acvf 0x0 C
acvf 0x8 C
!echo "Printing functions on offset 0x0 in all vtables"
acvf 0x0
!echo "Printing functions on offset 0x8 in all vtables"
acvf 0x8
!echo "Trying print out of bounds"
acvf 0x09 B
acvf 0x12312321 C
acvf 0x12312
```

```
B:
     0 vtable 0x3d50 @ +0x0 size:+0x10
     1 vtable 0x3d50 @ +0x0 size:+0x8
B
  (vtable at 0x3d50)
  (vtable at 0x3d50)
C
  (vtable at 0x3d30)
Printing functions on offset 0x0 and 0x8 in B vtable
Function address: 0x000011ca, in B vtable    0 
Function address: 0x000011ca, in B vtable    1 
Function address: 0x00001206, in B vtable    0 
Printing functions on offset 0x0 and 0x8 in C vtable
Function address: 0x00001242, in C vtable    0 
Function address: 0x00001206, in C vtable    0 
Printing functions on offset 0x0 in all vtables
Function address: 0x000011ca, in B vtable    0 
Function address: 0x000011ca, in B vtable    1 
Function address: 0x00001242, in C vtable    0 
Printing functions on offset 0x8 in all vtables
Function address: 0x00001206, in B vtable    0 
Function address: 0x00001206, in C vtable    0 
Trying print out of bounds
```

File: libr/core/cmd_anal.c [hunk 1]
```diff
@@ -9507,17 +9507,17 @@ static void cmd_anal_class_vtable(RCore *core, const char *input) {
...
+		char *class_arg = NULL;
 		if(end){
-			classNameArg = r_str_trim_head_ro(end);
+			class_arg = r_str_trim_head_ro(end);
 		}
 
-		if(classNameArg){
-			end = r_str_trim_head_wp(classNameArg); // in case of extra unwanted stuff at the cmd end
+		if(class_arg){
+			end = r_str_trim_head_wp(class_arg); // in case of extra unwanted stuff at the cmd end
```
Comment (radare) [inline libr/core/cmd_anal.c]: Missing space before (

File: libr/anal/class.c [hunk 2]
```diff
@@ -1163,4 +1163,53 @@ R_API void r_anal_class_list_vtables(RAnal *anal, const char *class_name) {
 		r_cons_printf ("  %4s vtable 0x%"PFMT64x" @ +0x%"PFMT64x" size:+0x%"PFMT64x"\n", vtable->id, vtable->addr, vtable->offset, vtable->size);
 	}
 	r_vector_free (vtables);
+}
+
+static void list_all_functions_at_vtable_offset(RAnal *anal, const char *class_name, ut64 offset){
+
```
Comment (radare) [inline libr/anal/class.c]: Empty line

File: libr/anal/class.c [hunk 3]
```diff
@@ -1163,4 +1163,53 @@ R_API void r_anal_class_list_vtables(RAnal *anal, const char *class_name) {
...
+	RVTableContext vtableContext;
+	r_anal_vtable_begin(anal, &vtableContext);
+	ut8 function_ptr_size = vtableContext.word_size; 
+
+	ut64 func_address;
+	RVector *vtables = r_anal_class_vtable_get_all (anal, class_name);
+	RAnalVTable *vtable;
+
+	if(!vtables) // no vtables
```
Comment (radare) [inline libr/anal/class.c]: Missing braces {}

File: libr/anal/class.c [hunk 4]
```diff
@@ -1163,4 +1163,53 @@ R_API void r_anal_class_list_vtables(RAnal *anal, const char *class_name) {
 		r_cons_printf ("  %4s vtable 0x%"PFMT64x" @ +0x%"PFMT64x" size:+0x%"PFMT64x"\n", vtable->id, vtable->addr, vtable->offset, vtable->size);
 	}
 	r_vector_free (vtables);
+}
+
+static void list_all_functions_at_vtable_offset(RAnal *anal, const char *class_name, ut64 offset){
```
Comment (radare) [inline libr/anal/class.c]: Missing space before {

File: libr/anal/class.c [hunk 5]
```diff
@@ -1163,4 +1163,53 @@ R_API void r_anal_class_list_vtables(RAnal *anal, const char *class_name) {
...
+
+	ut64 func_address;
+	RVector *vtables = r_anal_class_vtable_get_all (anal, class_name);
+	RAnalVTable *vtable;
+
+	if(!vtables) // no vtables
+		return;
+
+	r_vector_foreach (vtables, vtable) {
+		if(vtable->size < offset + function_ptr_size) {
```
Comment (radare) [inline libr/anal/class.c]: Missing space before (

File: libr/anal/class.c [hunk 6]
```diff
@@ -1163,4 +1163,53 @@ R_API void r_anal_class_list_vtables(RAnal *anal, const char *class_name) {
...
+			continue;
+		}
+
+		if(vtableContext.read_addr(anal, vtable->addr+offset, &func_address))
+			r_cons_printf("Function address: 0x%08"PFMT64x", in %s vtable %4s \n", func_address, class_name, vtable->id);
```
Comment (radare) [inline libr/anal/class.c]: Remove space at the end of string. Why use %4s instead of %s?

File: libr/anal/class.c [hunk 7]
```diff
@@ -1163,4 +1163,53 @@ R_API void r_anal_class_list_vtables(RAnal *anal, const char *class_name) {
...
+	}
+	r_vector_free (vtables);
+}
+
+R_API void r_anal_class_list_vtable_offset_functions(RAnal *anal, const char *class_name, ut64 offset){
+	if(!class_name) {
```
Comment (radare) [inline libr/anal/class.c]: Swap condition. Always true conditional blocks must go first

File: libr/anal/class.c [hunk 8]
```diff
@@ -1163,4 +1163,53 @@ R_API void r_anal_class_list_vtables(RAnal *anal, const char *class_name) {
...
+		SdbList *classes = r_anal_class_get_all (anal, true);
+		SdbListIter *iter;
+		SdbKv *kv;
+		ls_foreach (classes, iter, kv) {
+			char *class_name = sdbkv_key(kv);
```
Comment (radare) [inline libr/anal/class.c]: I think thats const. Otherwise its leaking

File: libr/anal/class.c [hunk 9]
```diff
@@ -1163,4 +1163,53 @@ R_API void r_anal_class_list_vtables(RAnal *anal, const char *class_name) {
...
+		if (!class_name_sanitized) {
+			return;
+		}
+		if (!r_anal_class_exists_raw (anal, class_name_sanitized)) {
+			free (class_name_sanitized);
+			return;
+		}
+		free (class_name_sanitized);
+
+		list_all_functions_at_vtable_offset(anal, class_name, offset);
```
Comment (radare) [inline libr/anal/class.c]: Missing space before (

File: libr/core/cmd_anal.c [hunk 10]
```diff
@@ -9492,6 +9492,36 @@ static void cmd_anal_class_vtable(RCore *core, const char *input) {
...
+			break;
+		}
+		char *end = strchr(cstr, ' ');
+		if (end) {
+			*end = '\0';
+			end++;
+		}
+		ut64 offset_arg = r_num_get(core->num, cstr);
+		char *class_arg = NULL;
+		if(end){
```
Comment (radare) [inline libr/core/cmd_anal.c]: Missing space before {

File: libr/anal/class.c (hunk 8 above)
Comment (HoundThe) [reply libr/anal/class.c]: I peeked the sdbkv_key() function, it doesn't allocate anything, but changed it to const

File: libr/anal/class.c (hunk 6 above)
Comment (HoundThe) [reply libr/anal/class.c]: I saw the r_anal_class_list_vtables() prints it that way, so I wanted to stay consistent

**Comment by HoundThe** [comment]
Thanks for the feedback! Could you please elaborate on the * suffix? I'll fix the rest when I'll be home

File: libr/anal/class.c (hunk 7 above)
Comment (HoundThe) [reply libr/anal/class.c]: I am not sure what do you mean, there isn't any always true block there, there are 2 options based on class_name

Comment (radare) [reply libr/anal/class.c]: Swap the conditi0n. Use if (class_name)

**Comment by XVilka** [comment]
LGTM.

**Comment by thestr4ng3r** [comment]
Looks good, please add some tests.

**Comment by radare** [comment]
lgtm, add one test at least we'll merge. thanks!!

**Comment by HoundThe** [comment]
Sorry for the delay, I've added a test, used already existing binary that is used for av tests. Got a question, should I allow entering negative offset? Also should I add the * suffix?

**Pull request merged**